File: Samples/WindowsML/Shared/cs/ImageProcessor.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.md in the repo root for license information.

using Microsoft.ML.OnnxRuntime.Tensors;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using Windows.Graphics.Imaging;
using Windows.Media;
using Windows.Storage;
using Windows.Storage.Streams;

namespace WindowsML.Shared
{
    /// <summary>
    /// Image loading and preprocessing for ML inference
    /// </summary>
    public static class ImageProcessor
    {
        /// <summary>
        /// Load image file as VideoFrame
        /// </summary>
        public static async Task<VideoFrame> LoadImageFileAsync(string filePath)
        {
            try
            {
                StorageFile file = await StorageFile.GetFileFromPathAsync(filePath);
                IRandomAccessStream stream = await file.OpenAsync(FileAccessMode.Read);
                BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);
                SoftwareBitmap softwareBitmap = await decoder.GetSoftwareBitmapAsync();
                VideoFrame inputImage = VideoFrame.CreateWithSoftwareBitmap(softwareBitmap);

                return inputImage;
            }
            catch (FileNotFoundException)
            {
                throw new FileNotFoundException($"Image file not found: {filePath}");
            }
            catch (Exception ex)
            {
                throw new Exception($"Error loading image: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Preprocess image for ML inference
        /// </summary>
        public static async Task<DenseTensor<float>> PreprocessImageAsync(VideoFrame videoFrame)
        {
            SoftwareBitmap softwareBitmap = videoFrame.SoftwareBitmap;
            const int targetWidth = 224;
            const int targetHeight = 224;

            float[] mean = [0.485f, 0.456f, 0.406f];
            float[] std = [0.229f, 0.224f, 0.225f];

            // Convert to BGRA8
            if (softwareBitmap.BitmapPixelFormat != BitmapPixelFormat.Bgra8 ||
                softwareBitmap.BitmapAlphaMode != BitmapAlphaMode.Premultiplied)
            {
                softwareBitmap = SoftwareBitmap.Convert(softwareBitmap, BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied);
            }

            // Resize
            softwareBitmap = await ResizeSoftwareBitmapAsync(softwareBitmap, targetWidth, targetHeight);

            // Get pixel data
            uint bufferSize = (uint)(targetWidth * targetHeight * 4);
            Windows.Storage.Streams.Buffer buffer = new(bufferSize);
            softwareBitmap.CopyToBuffer(buffer);
            byte[] pixels = buffer.ToArray();

            // Output Tensor shape: [1, 3, 224, 224]
            DenseTensor<float> tensorData = new([1, 3, targetHeight, targetWidth]);

            for (int y = 0; y < targetHeight; y++)
            {
                for (int x = 0; x < targetWidth; x++)
                {
                    int pixelIndex = (y * targetWidth + x) * 4;
                    float r = pixels[pixelIndex + 2] / 255f;
                    float g = pixels[pixelIndex + 1] / 255f;
                    float b = pixels[pixelIndex + 0] / 255f;

                    // Normalize using mean/stddev
                    r = (r - mean[0]) / std[0];
                    g = (g - mean[1]) / std[1];
                    b = (b - mean[2]) / std[2];

                    int baseIndex = y * targetWidth + x;
                    tensorData[0, 0, y, x] = r; // R
                    tensorData[0, 1, y, x] = g; // G
                    tensorData[0, 2, y, x] = b; // B
                }
            }

            return tensorData;
        }

        /// <summary>
        /// Resize SoftwareBitmap to specified dimensions
        /// </summary>
        private static async Task<SoftwareBitmap> ResizeSoftwareBitmapAsync(SoftwareBitmap bitmap, int width, int height)
        {
            using InMemoryRandomAccessStream stream = new();
            BitmapEncoder encoder = await BitmapEncoder.CreateAsync(BitmapEncoder.PngEncoderId, stream);
            encoder.SetSoftwareBitmap(bitmap);
            encoder.IsThumbnailGenerated = false;
            await encoder.FlushAsync();
            stream.Seek(0);

            BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);
            BitmapTransform transform = new() {
                ScaledWidth = (uint)width,
                ScaledHeight = (uint)height,
                InterpolationMode = BitmapInterpolationMode.Fant
            };
            SoftwareBitmap resized = await decoder.GetSoftwareBitmapAsync(
                BitmapPixelFormat.Bgra8,
                BitmapAlphaMode.Premultiplied,
                transform,
                ExifOrientationMode.IgnoreExifOrientation,
                ColorManagementMode.DoNotColorManage);

            stream.Dispose();
            return resized;
        }
    }
}

File: Samples/WindowsML/Shared/cpp/ImageProcessor.cpp
// Copyright (c) Microsoft Corporation. All rights r        auto spByteAccess = reference.as<::Windows::Foundation::IMemoryBufferByteAccess>();served.
// Licensed under the MIT License. See LICENSE.md in the repo root for license information.
#include "ImageProcessor.h"
#include <MemoryBuffer.h>
#include <winrt/Windows.Storage.h>
#include <winrt/Windows.Storage.Streams.h>
#include <winrt/Windows.Graphics.Imaging.h>
#include <winrt/Windows.Foundation.h>

using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Graphics::Imaging;
using namespace winrt::Windows::Media;
using namespace winrt::Windows::Storage;
using namespace winrt::Windows::Storage::Streams;

namespace WindowsML
{
namespace Shared
{

    IAsyncOperation<VideoFrame> ImageProcessor::LoadImageFileAsync(winrt::hstring filePath)
    {
        StorageFile file = co_await StorageFile::GetFileFromPathAsync(filePath);
        auto stream = co_await file.OpenAsync(FileAccessMode::Read);
        BitmapDecoder decoder = co_await BitmapDecoder::CreateAsync(stream);
        SoftwareBitmap softwareBitmap = co_await decoder.GetSoftwareBitmapAsync();
        VideoFrame inputImage = VideoFrame::CreateWithSoftwareBitmap(softwareBitmap);

        co_return inputImage;
    }

    std::vector<float> ImageProcessor::BindVideoFrameAsTensor(const VideoFrame& frame)
    {
        SoftwareBitmap bitmap = frame.SoftwareBitmap();
        SoftwareBitmap bitmapBgra8 = SoftwareBitmap::Convert(bitmap, BitmapPixelFormat::Bgra8, BitmapAlphaMode::Ignore);

        InMemoryRandomAccessStream stream = InMemoryRandomAccessStream();

        BitmapEncoder encoder = BitmapEncoder::CreateAsync(BitmapEncoder::BmpEncoderId(), stream).get();
        encoder.SetSoftwareBitmap(bitmapBgra8);
        encoder.BitmapTransform().ScaledWidth(IMAGE_SIZE);
        encoder.BitmapTransform().ScaledHeight(IMAGE_SIZE);
        encoder.BitmapTransform().InterpolationMode(BitmapInterpolationMode::Fant);
        encoder.FlushAsync().get();

        stream.Seek(0);
        BitmapDecoder decoder = BitmapDecoder::CreateAsync(stream).get();
        SoftwareBitmap resizedBitmap = decoder.GetSoftwareBitmapAsync(BitmapPixelFormat::Bgra8, BitmapAlphaMode::Ignore).get();

        BitmapBuffer bitmapBuffer = resizedBitmap.LockBuffer(BitmapBufferAccessMode::Read);

        IMemoryBufferReference reference = bitmapBuffer.CreateReference();

        auto spByteAccess = reference.as<::Windows::Foundation::IMemoryBufferByteAccess>();
        uint8_t* pixelData = nullptr;
        uint32_t pixelDataCapacity = 0;

        spByteAccess->GetBuffer(&pixelData, &pixelDataCapacity);

        std::vector<float> tensorData(CHANNELS * IMAGE_SIZE * IMAGE_SIZE);

        for (int y = 0; y < IMAGE_SIZE; ++y)
        {
            for (int x = 0; x < IMAGE_SIZE; ++x)
            {
                int idx = (y * IMAGE_SIZE + x) * 4; // BGRA stride
                int offset = y * IMAGE_SIZE + x;

                // The required input size is (3x224x224),
                // normalized using mean=[0.485, 0.456, 0.406]
                // and std=[0.229, 0.224, 0.225]
                float r = static_cast<float>(pixelData[idx + 2]) / 255.0f;
                float g = static_cast<float>(pixelData[idx + 1]) / 255.0f;
                float b = static_cast<float>(pixelData[idx + 0]) / 255.0f;

                tensorData[0 * IMAGE_SIZE * IMAGE_SIZE + offset] = (r - MEAN_R) / STD_R;
                tensorData[1 * IMAGE_SIZE * IMAGE_SIZE + offset] = (g - MEAN_G) / STD_G;
                tensorData[2 * IMAGE_SIZE * IMAGE_SIZE + offset] = (b - MEAN_B) / STD_B;
            }
        }

        return tensorData;
    }

} // namespace Shared
} // namespace WindowsML
